src/classpath-openjdk.cpp

```diff
@@ -975,6 +978,22 @@ class EmbeddedFile {
   unsigned pathLength;
 };
 
+void uncaughtException(Thread *t, GcThrowable *e)
+{
+  GcMethod* dispatch = resolveMethod(t,
+                                     roots(t)->bootLoader(),
+                                     "java/lang/Thread",
+                                     "dispatchUncaughtException",
+                                     "(Ljava/lang/Throwable;)V");
+  if (dispatch != NULL) {
+    t->m->processor->invoke(t, dispatch, t->javaThread, e);
+
+    // We ignore any exceptions from the uncaught
+    // exception handler itself.
+    t->exception = NULL;
```

Comment:
I don't think this is enough, since the invoke call above may unwind the stack past it.  Have you tested an exception handler that throws an exception?  I recommend adding such a test to ThreadExceptions.java.  We might need to add a THREAD_RESOURCE0 here to catch the exception and clear it.

Reply:
e.g. something like this:

```c++
THREAD_RESOURCE0(t, {
  t->exception = NULL;
});

t->m->processor->invoke(t, dispatch, t->javaThread, e);
```

Reply:
I see your point: when the `uncaughtExceptionHandler` throws an exception, the `t-checkpoint->unwind()` statement aborts the release of the previous `THREAD_RESOURCE0`. It looks like this might need a custom `Checkpoint` object. Or wrap the method inside a Java `try-catch` block.